### PR TITLE
get a list of expected layers before the IF is finished

### PR DESCRIPTION
**Pull request by Gustry**
### What does it fix?
* Ticket: #4495
* Funded by: DFAT
* Description: Just before my flight, I tried to do the request we talked this afternoon.

This is WIP, I need to take off ! ;-)

### Checklist:
- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Add to the changelog in metadata.txt if it's a new feature
- [ ] Unit test for new code added
- [ ] Request someone to review or test your PR

### CODE REVIEW

File: safe/impact_function/impact_function.py [hunk 1]
```diff
@@ -966,6 +972,33 @@ def prepare(self):
...
+        """
+        if not self._is_ready:
+            return None
+        else:
+            return self._output_number()
+
+    def _output_number(self):
+        """Compute the number of outputs that the IF will produce.
+
+        Be careful when you call this function.Better to use `output_number()`.
```
Comment (ismailsunni) [inline safe/impact_function/impact_function.py]: Space after `.`

File: safe/impact_function/impact_function.py [hunk 2]
```diff
@@ -966,6 +972,33 @@ def prepare(self):
...
+
+        :return: The number of output.
+        :rtype: int
+        """
+        if is_raster_layer(self.exposure):
+            if self.exposure.keywords.get('layer_mode') == 'continuous':
+                return 4
```
Comment (ismailsunni) [inline safe/impact_function/impact_function.py]: Can you give comment what does 4 or 5 mean? Or where is it come from? Perhaps listing the output layers in the method's docstring is good for the other developers.

File: safe/impact_function/test/test_impact_function.py [hunk 3]
```diff
@@ -627,7 +634,8 @@ def test_scenario(self, scenario_path=None):
         LOGGER.info('Running the scenario : %s' % scenario_path)
         scenario, expected_steps, expected_outputs = read_json_flow(
             scenario_path)
-        status, steps, outputs = run_scenario(scenario, use_debug)
+        status, steps, outputs, computed_nb_outputs = \
```
Comment (ismailsunni) [inline safe/impact_function/test/test_impact_function.py]: Remove the backslash? and put new line after the opening bracket.

File: safe/impact_function/impact_function.py [hunk 4]
```diff
@@ -436,6 +436,12 @@ def outputs(self):
 
         # Remove layers which are not set.
         layers = [layer for layer in layers.values() if layer]
+
+        if len(layers) != self._output_number():
+            # If this happen, it's an error from developpers.
+            raise Exception(
```
Comment (ismailsunni) [inline safe/impact_function/impact_function.py]: Perhaps give more information, like how many expected layers and what it gets, to see where is the problem.
Even better, if we can also show which layer is missing (but it seems your current code doesn't support this).

**Comment by Gustry** [comment]
THanks, it was an early review. I just made a quick PR before my flight on Friday evening.

File: safe/impact_function/impact_function.py (hunk 1 above)
Comment (Gustry) [reply safe/impact_function/impact_function.py]: I guess it's because of PEP8 :D That's why. Ok, I will create a new line

**Comment by Gustry** [comment]
Ok, this PR is ready to be reviewed.
@myarjunar After this PR is merged, you can add the dialog to ask which layer you want to be printed in the report.

You use it like this:
```
impact_function.prepare()
list_of_expected_layers = impact_function.output_layers_expected() # it will return a list of layer_purpose that the IF will give at the end
impact_function.run()
```

**Pull request merged**